### PR TITLE
Adding the extension methods for FastTree classification and regression

### PR DESCRIPTION
**Pull request by sfilipi**
OnGoing work to address #754. 

Added the extensions to the Binary and Regression context for the respective FastTree trainers. 
Ranking coming on a separate PR, since it needs a bit more work with the context and elevator changes.

### CODE REVIEW

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            Contracts.CheckValue(label, nameof(label));
+            Contracts.CheckValue(features, nameof(features));
+            Contracts.CheckValueOrNull(weights);
+            Contracts.CheckParam(numLeaves >= 2, nameof(numLeaves), "Must be at least 2.");
+            Contracts.CheckParam(numTrees > 0, nameof(numTrees), "Must be positive");
+            Contracts.CheckParam(minDocumentsInLeafs > 0, nameof(minDocumentsInLeafs), "Must be positive");
+            Contracts.CheckParam(learningRates > 0, nameof(learningRates), "Must be positive");
+            Contracts.CheckValueOrNull(advancedSettings);
+            Contracts.CheckValueOrNull(onFit);
+
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTreeStatic.cs]: don't duplicate

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs [hunk 2]
```diff
@@ -120,18 +120,46 @@ public sealed partial class FastTreeBinaryClassificationTrainer :
...
+            {
+                //take a quick snapshot at the defaults, for comparison with the current args values
+                var snapshot = new Arguments();
+
+                // Check that the user didn't supply different parameters in the args, from what it specified directly.
+                TrainerUtils.CheckArgsAndAdvancedSettingMismatch(numLeaves, snapshot.NumLeaves, Args.NumLeaves, nameof(numLeaves));
+                TrainerUtils.CheckArgsAndAdvancedSettingMismatch(numTrees, snapshot.NumTrees, Args.NumTrees, nameof(numTrees));
+                TrainerUtils.CheckArgsAndAdvancedSettingMismatch(minDocumentsInLeafs, snapshot.MinDocumentsInLeafs, Args.MinDocumentsInLeafs, nameof(minDocumentsInLeafs));
+                TrainerUtils.CheckArgsAndAdvancedSettingMismatch(learningRates, snapshot.LearningRates, Args.LearningRates, nameof(learningRates));
+            }
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTreeClassification.cs]: move in FastTreeUtils and re-use.

File: src/Microsoft.ML.Data/Training/TrainerUtils.cs [hunk 3]
```diff
@@ -382,6 +383,24 @@ public static SchemaShape.Column MakeR4ScalarWeightColumn(string weightColumn)
...
+            if (argValue != defaultColName)
+                throw Contracts.Except($"Don't supply a value for the {defaultColName} column in the arguments, as it will be ignored. Specify them in the loader, or constructor instead instead.");
+        }
+
+        public static void CheckArgsAndAdvancedSettingMismatch<T>(T methodParam, T defaultVal, T setting, string argName)
+        {
+            // if, after applying the advancedArgs delegate, the args are different that the default value
+            // and are also different than the value supplied directly to the xtension method, warn the user.
+            if (!setting.Equals(defaultVal) && !setting.Equals(methodParam))
+                Console.WriteLine($"The value supplied to advanced settings , is different than the value supplied directly. Using value {setting} for {argName}");
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Training/TrainerUtils.cs]: > Console.WriteLine [](start = 15, length = 18)

I don't like this. Can we accept Host or Channel as first object and use it?

File: src/Microsoft.ML.Data/Training/TrainerUtils.cs [hunk 4]
```diff
@@ -382,6 +383,24 @@ public static SchemaShape.Column MakeR4ScalarWeightColumn(string weightColumn)
...
         }
+
+        /// <summary>
+        /// Check that the label, feature, weights is not supplied in the args the constructor.
+        /// Those parameters should be internal if they are not used from the maml help code path.
+        /// </summary>
+        public static void CheckArgumesDefaultColNames(string defaultColName, string argValue)
+        {
+            if (argValue != defaultColName)
+                throw Contracts.Except($"Don't supply a value for the {defaultColName} column in the arguments, as it will be ignored. Specify them in the loader, or constructor instead instead.");
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Training/TrainerUtils.cs]: Everywhere you call this you have host or environment. Why you don't want to use it?

File: src/Microsoft.ML.FastTree/FastTree.cs [hunk 5]
```diff
@@ -99,6 +99,13 @@ public abstract class FastTreeTrainerBase<TArgs, TTransformer, TModel> :
 
             //apply the advanced args, if the user supplied any
             advancedSettings?.Invoke(Args);
+
+            // check that the users didn't specify different label, group, feature, weights in the args, from what they supplied directly
+            TrainerUtils.CheckArgumesDefaultColNames(DefaultColumnNames.Label, Args.LabelColumn);
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/FastTree.cs]: >Argumes [](start = 30, length = 7)

typo

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs [hunk 6]
```diff
@@ -120,18 +120,46 @@ public sealed partial class FastTreeBinaryClassificationTrainer :
...
+        /// <param name="numLeaves">The max number of leaves in each regression tree.</param>
+        /// <param name="numTrees">Total number of decision trees to create in the ensemble.</param>
+        public FastTreeBinaryClassificationTrainer(IHostEnvironment env,
+            string labelColumn,
+            string featureColumn,
+            string weightColumn = null,
+            int numLeaves = 20,
+            int numTrees = 100,
+            int minDocumentsInLeafs = 10,
+            double learningRates = 0.2,
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/FastTreeClassification.cs]: >learningRates [](start = 19, length = 13)

just one learning rate will suffice

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs [hunk 7]
```diff
@@ -120,18 +120,46 @@ public sealed partial class FastTreeBinaryClassificationTrainer :
...
             // Set the sigmoid parameter to the 2 * learning rate, for traditional FastTreeClassification loss
             _sigmoidParameter = 2.0 * Args.LearningRates;
+
+            if (advancedSettings != null)
+            {
+                //take a quick snapshot at the defaults, for comparison with the current args values
+                var snapshot = new Arguments();
+
+                // Check that the user didn't supply different parameters in the args, from what it specified directly.
+                TrainerUtils.CheckArgsAndAdvancedSettingMismatch(numLeaves, snapshot.NumLeaves, Args.NumLeaves, nameof(numLeaves));
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/FastTreeClassification.cs]: >CheckArgsAndAdvancedSettingMismatch [](start = 29, length = 35)

actually this is not necessary. If they are so cool to set the values in 2 places, that's fine.
I wanted it to be checked on columns, since these are ignored. The other args, you set them first, then you invoke delegate and you're good.

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        public static (Scalar<float> score, Scalar<float> probability, Scalar<bool> predictedLabel) FastTree(this BinaryClassificationContext.BinaryClassificationTrainers ctx,
+            Scalar<bool> label, Vector<float> features, Scalar<float> weights = null,
+            int numLeaves = 20,
+            int numTrees = 100,
+            int minDocumentsInLeafs = 10,
+            double learningRates = 0.2,
+            Action<FastTreeBinaryClassificationTrainer.Arguments> advancedSettings = null,
+            Action<IPredictorWithFeatureWeights<float>> onFit = null)
```
Comment (Zruty0) [inline src/Microsoft.ML.FastTree/FastTreeStatic.cs]: >IPredictorWithFeatureWeights [](start = 19, length = 28)

that's the best we could do here? I would argue that we should expose some form of tree ensemble predictor or something

File: src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs [hunk 9]
```diff
@@ -59,10 +59,10 @@ public static class FactorizationMachineStatic
                 var trainer = new FieldAwareFactorizationMachineTrainer(env, labelCol, featureCols, advancedSettings:
                     args =>
                     {
+                        advancedSettings?.Invoke(args);
```
Comment (Zruty0) [inline src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineStatic.cs]: >advancedSettings [](start = 24, length = 16)

my order was intentional :)

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+using Microsoft.ML.Runtime.Training;
+using System;
+
+namespace Microsoft.ML.Trainers
+{
+    public static class FastTreeStatic
+    {
+        public static Scalar<float> FastTree(this RegressionContext.RegressionTrainers ctx,
+            Scalar<float> label, Vector<float> features, Scalar<float> weights = null,
+            int numLeaves = 20,
```
[the remark this answers is not in the document]
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTreeStatic.cs]: Are you asking why defaults? To give people an idea of a good start, if they wanted to approximate.. 
To avoid dealing with nulls. those are going to be the values if not set to anything. 



---
In reply to: [220018895](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220018895) [](ancestors = 220018895)

File: src/Microsoft.ML.Data/Training/TrainerUtils.cs (hunk 3 above)
Comment (sfilipi) [reply src/Microsoft.ML.Data/Training/TrainerUtils.cs]: for sure; put it to look for a Warn method, or pass the channel, and forgot to replace. 

---
In reply to: [220018209](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220018209) [](ancestors = 220018209)

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs (hunk 8 above)
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTreeStatic.cs]: I'll leave it for now and log an issue about it. Just casting didn't work. 

---
In reply to: [220030085](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220030085) [](ancestors = 220030085)

File: src/Microsoft.ML.FastTree/FastTreeClassification.cs (hunk 7 above)
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTreeClassification.cs]: well,  this is trying to catch the case where they set one value in one place, and another one in the other. No exceptions, just Warn them about which value will be used. 

---
In reply to: [220029873](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220029873) [](ancestors = 220029873)

File: src/Microsoft.ML.FastTree/FastTreeRegression.cs [hunk 11]
```diff
@@ -57,15 +57,37 @@ public sealed partial class FastTreeRegressionTrainer
...
-            : base(env, TrainerUtils.MakeR4ScalarLabel(labelColumn), featureColumn, weightColumn, groupIdColumn, advancedSettings)
+        /// <param name="learningRates">The learning rate.</param>
+        /// <param name="minDocumentsInLeafs">The minimal number of documents allowed in a leaf of a regression tree, out of the subsampled data.</param>
+        /// <param name="numLeaves">The max number of leaves in each regression tree.</param>
+        /// <param name="numTrees">Total number of decision trees to create in the ensemble.</param>
+        public FastTreeRegressionTrainer(IHostEnvironment env,
+            string labelColumn,
+            string featureColumn,
+            string weightColumn = null,
+            int numLeaves = 20,
```
[the remark this answers is not in the document]
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTreeRegression.cs]: Thanks. 

---
In reply to: [220018807](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220018807) [](ancestors = 220018807)

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+               }, label, features, weights);
+
+            return rec.Output;
+        }
+
+        private static void CheckUserValues<TVal, TArgs, TPred>(Scalar<TVal> label, Vector<float> features, Scalar<float> weights = null,
+            int numLeaves = Defaults.NumLeaves,
+            int numTrees = Defaults.NumTrees,
+            int minDocumentsInLeafs = Defaults.MinDocumentsInLeafs,
+            double learningRate = Defaults.LearningRates,
```
Comment (sfilipi) [inline src/Microsoft.ML.FastTree/FastTreeStatic.cs]: fixxxx

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj [hunk 13]
```diff
@@ -7,13 +7,14 @@
...
-    <NativeAssemblyReference Include="FactorizationMachineNative" />
-
+    
     <ProjectReference Include="..\..\src\Microsoft.ML.Analyzer\Microsoft.ML.Analyzer.csproj">
       <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
       <OutputItemType>Analyzer</OutputItemType>
     </ProjectReference>
 
     <NativeAssemblyReference Include="CpuMathNative" />
+    <NativeAssemblyReference Include="FactorizationMachineNative" />
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: I can see tests for FastTree added, but nothing regarding FM

File: src/Microsoft.ML.FastTree/FastTreeStatic.cs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+using Microsoft.ML.Runtime.FastTree;
+using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Runtime.Training;
+using System;
+
+namespace Microsoft.ML.Trainers
+{
+    public static class FastTreeStatic
+    {
+        public static Scalar<float> FastTree(this RegressionContext.RegressionTrainers ctx,
```
[the remark this answers is not in the document]
Comment (sfilipi) [reply src/Microsoft.ML.FastTree/FastTreeStatic.cs]: I did them, before you commented!!!! see the next push :)

---
In reply to: [220283064](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220283064) [](ancestors = 220283064)

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj (hunk 13 above)
Comment (sfilipi) [reply test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: it was already there. I just moved the reference a few lines down, to keep it together with the other native references. 

---
In reply to: [220283652](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220283652) [](ancestors = 220283652)

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj [hunk 15]
```diff
@@ -6,15 +6,15 @@
...
     <ProjectReference Include="..\..\src\Microsoft.ML.Analyzer\Microsoft.ML.Analyzer.csproj">
       <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
       <OutputItemType>Analyzer</OutputItemType>
     </ProjectReference>
 
     <NativeAssemblyReference Include="CpuMathNative" />
-    <NativeAssemblyReference Include="LdaNative" />
+    <NativeAssemblyReference Include="FactorizationMachineNative" />
+    <NativeAssemblyReference Include="FastTreeNative" />
+	<NativeAssemblyReference Include="LdaNative" />
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: >	 [](start = 0, length = 1)

tab

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj [hunk 16]
```diff
@@ -6,15 +6,15 @@
     <ProjectReference Include="..\..\src\Microsoft.ML.Data\Microsoft.ML.Data.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.ImageAnalytics\Microsoft.ML.ImageAnalytics.csproj" />
     <ProjectReference Include="..\..\src\Microsoft.ML.StandardLearners\Microsoft.ML.StandardLearners.csproj" />
-    <ProjectReference Include="..\Microsoft.ML.TestFramework\Microsoft.ML.TestFramework.csproj" />
-    <NativeAssemblyReference Include="FactorizationMachineNative" />
-
+    <ProjectReference Include="..\Microsoft.ML.TestFramework\Microsoft.ML.TestFramework.csproj" />
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: >     [](start = 98, length = 4)

is...
this...
a...
EXTRA SPACES?

File: test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj (hunk 15 above)
Comment (sfilipi) [reply test/Microsoft.ML.StaticPipelineTesting/Microsoft.ML.StaticPipelineTesting.csproj]: Tools->Options->Text Editor->All Languages->Insert Spaces

Checked!

---
In reply to: [220301837](https://github.com/dotnet/machinelearning/pull/1009#discussion_r220301837) [](ancestors = 220301837)
